README.md
# Cherow

[![NPM version](https://img.shields.io/npm/v/cherow.svg)](https://www.npmjs.com/package/cherow)
[![Gitter chat](https://badges.gitter.im/gitterHQ/gitter.png)](https://gitter.im/cherow/cherow)
[![Build Status](https://travis-ci.org/cherow/cherow.svg?branch=master)](https://travis-ci.org/cherow/cherow)
[![Build status](https://ci.appveyor.com/api/projects/status/pkydnkv0deckns5l/branch/master?svg=true)](https://ci.appveyor.com/project/Kflash/cherow/branch/master)
[![CircleCI](https://circleci.com/gh/cherow/cherow.svg?style=svg)](https://circleci.com/gh/cherow/cherow)
[![Coverage Status](https://coveralls.io/repos/github/cherow/cherow/badge.svg?branch=master)](https://coveralls.io/github/cherow/cherow?branch=master)

A very fast, standards-compliant, self-hosted javascript parser with high focus on both performance and stability.

It strictly follows the [ECMAScript® 2018 Language Specification](https://tc39.github.io/ecma262/) and should parse according to these specifications.

## [Demo](https://cherow.github.io/cherow/) and [Benchmark](https://cherow.github.io/cherow/performance/)

## Features

* Full support for ECMAScript® 2018 [(ECMA-262 9th Edition)](https://tc39.github.io/ecma262/) (*latest draft*)
* Optimized for handheld devices 
* Stage 3 (*ESNext*) proposals via option
* [JSX](https://reactjs.org/docs/jsx-in-depth.html), a syntax extension for React
* Skips hashbang comment nodes by default
* Skips BOM (*U+FEFF*) by default
* Tolerant parsing
* Optional tracking of syntax node location (index-based and line-column)
* Emits an [ESTree-compatible](https://github.com/estree/estree) abstract syntax tree.
* Very well tested (~55 000 [unit tests](https://github.com/cherow/cherow/tree/master/test) with [full code coverage)](https://coveralls.io/github/cherow/cherow))
* 97% Test262 Test coverage
* 30 - 45% faster than [Acorn](https://github.com/acornjs/acorn) and [Esprima](https://github.com/jquery/esprima)

## ESNext features

`Stage 3` features support. These need to be enabled with the `next` option. 

* [Import()](https://github.com/tc39/proposal-dynamic-import)
* [Class field declarations for JavaScript](https://github.com/tc39/proposal-class-fields)
* [Numeric Separators](https://github.com/tc39/proposal-numeric-separator)
* [Private methods and getter/setters for JavaScript classes](https://github.com/tc39/proposal-private-methods)
* [Optional catch binding](https://github.com/tc39/proposal-optional-catch-binding)
* [JSON Superset](https://github.com/tc39/proposal-json-superset)
* [BigInt](https://github.com/tc39/proposal-bigint)
* [Import.meta](https://github.com/tc39/proposal-import-meta)

## API

Cherow generates AST according to [ESTree AST format](https://github.com/estree/estree), and can be used to perform [syntactic analysis](https://en.wikipedia.org/wiki/Parsing) (parsing) of a JavaScript program, and with ES2015 and later a JavaScript program can be either [a script or a module](http://www.ecma-international.org/ecma-262/8.0/index.html#sec-ecmascript-language-scripts-and-modules) and this is achieved by choosing [`parseScript`](http://www.ecma-international.org/ecma-262/8.0/#sec-parse-script) function to parse a script and [`parseModule`](http://www.ecma-international.org/ecma-262/8.0/#sec-parsemodule) function to parse a module.

Here is a quick example:

```js

cherow.parseScript('const fooBar = 123;', { ranges: true});

```

This will return when serialized in json:

```js
{
    "type": "Program",
    "sourceType": "script",
    "body": [
        {
            "type": "VariableDeclaration",
            "declarations": [
                {
                    "type": "VariableDeclarator",
                    "init": {
                        "type": "Literal",
                        "value": 123,
                        "start": 15,
                        "end": 18
                    },
                    "id": {
                        "type": "Identifier",
                        "name": "fooBar",
                        "start": 6,
                        "end": 12
                    },
                    "start": 6,
                    "end": 18
                }
            ],
            "kind": "const",
            "start": 0,
            "end": 19
        }
    ],
    "start": 0,
    "end": 19
}
```

## Options

There is a second argument to both methods that allows you to specify various options:

| Option        | Description |
| ----------- | ------------------------------------------------------------ |
| `comments`        | Create a top-level comments array containing all comments |
| `tolerant`        | Create a top-level error array containing all "skipped" errors |
| `delegate`        | Accepts a callback function to be invoked for each syntax node (as the node is constructed) |
| `jsx`             | Enable React JSX parsing |
| `loc      `       | Attach line/column location information to each node |
| `ranges`          | Append start and end offsets to each node |
| `impliedStrict`   | Enable implied strict mode |
| `next`            | Enable stage 3 support (*ESNext*)  |
| `plugins`         | Array containing the plugins that you want to enable   |
| `source`          | Set to true to record the source file in every node's `loc` object when the `loc option` is set.  |
| `raw`             | Attach raw property to each literal node    |

## Comments

Single line, multiline and HTML comments are supported by `Cherow`, and the parser can be instructed to collect comments by setting the `comments option` to *true*.

A top-level comments array containing all comments will be attached to the root node (*Program*), and the type of each comment can 
either be [`SingleLine`](https://tc39.github.io/ecma262/#prod-SingleLineComment) for a single-line comment (`//`) or [`MultiLine`](https://tc39.github.io/ecma262/#prod-MultiLineComment) for a MultiLineComment (`/* */`).

HTML comments is not a part of the ECMAScript specifications, and the way Cherow deals with them deviates slightly from other
parsers. In Cherow `HTMLOpen` are used for a HTML open comments (`<!--`) and `HTMLClose` for a HTML close comment (`-->`).
In other  ECMAScripts parsers both are seen as a [`single-line comment `](https://tc39.github.io/ecma262/#prod-SingleLineComment).

# Tolerant parsing

The tolerant algorithm used by Cherow deviates slightly from both `Esprima` and Acorn due to the parsers complexity, and it's primarily for early errors, and other errors that are basically valid syntax but just not allowed.

A top-level errors array containing all "skipped" errors will be attached to the root node (Program),

## Plugins

Cherow can be extended through plugins. See [this repo](https://github.com/cherow/cherow-do-expressions) as an example on how to build your own plugins. 

By default Cherow pass around context masks everywhere as a simple immutable bit set, and also let you use mutable parser flags, in 
case any flags need passed by reference. You can either create your own masks or adopt the build in one as you find in the `do-expressions-repo`
or in the source folder on this repo.

**To create and add a plugin:**

```js

 // Create the do-expression plugin
export default function(Parser) {

    return class extends Parser {

        parsePrimaryExpression(context, pos) {
            return this.token === Token.DoKeyword 
            ? this.parseDoExpression(context)
            : super.parsePrimaryExpression(context, pos);

        }

        parseDoExpression(context) {
            const pos = this.getLocation();
            this.expect(context, Token.DoKeyword);
            const body = this.parseBlockStatement(context);
            return this.finishNode(context, pos, {
                type: 'DoExpression',
                body
            });
        }
     }
}

//  Use the plugin
Cherow.parseScript("let x = do {}", { plugins: [doExpressions] }
```

### Existing plugins


| Name        | Description |
| ----------- | ------------------------------------------------------------ |
| [`cherow-do-expression`](https://github.com/cherow/cherow-do-expressions)  | Stage 1 proposal |


## Walk / Syntax Delegate

The `delegate` option accepts a callback function to be invoked for each syntax node (as the node is constructed). 

Both ESTree AST nodes and comments can be delegated.

An advanced example can be seen [here](https://github.com/cherow/cherow/blob/master/test/miscellaneous/delegate.ts#L103)

And here is an quick example:

```js
    cherow.parseScript('foo // comment', { delegate: function(node) { console.log(node) } } )
```

This will output:

```js
    { type: 'SingleLine', value: ' comment', start: 4, end: 14 }
    { type: 'Identifier', name: 'foo' }
    { type: 'ExpressionStatement', expression: { type: 'Identifier', name: 'foo' } }
```        

## Bug reporting

If you caught a bug, don't hesitate to report it in the issue tracker. From the moment I respond to you, it will take maximum 60 minutes before the bug is fixed. 
Note that I will try to respond to you within one hour. Sometimes it can take a bit longer. I'm not allways online. And if I find out it 
will take more then 60 minutes to solve your issue, you will be notified. 

I know how irritating it can be if you are writing code and encounter bugs in your dependencies. And even more frustrating if you need to wait weeks or days.

## Contribution

If you feel something could've been done better, please do feel free to file a pull request with the changes.

Read our guidelines [here](CONTRIBUTING.md)


CHANGELOG.MD
# Changelog

*2018-7-7: Version 1.0.10*

 * Added support for JSX parsing through plugins

*2018-2-23: Version 1.2.0*

 * Improved performance
 * Simplified label statement parsing
 * Re-factor arrow function parsing

*2018-2-23: Version 1.2.1*

 * Improved performance
 * Fixed a couple of slip-ups
 
 *2018-3.5: Version 1.2.8

 * Added options for walk / syntax delegate as third argument
 * Exposed *.TSD files

*2018-3.7: Version 1.3.0

 * Refactored and optimized number scanning
 * Fixed several issues with number scanning (*decimals with leading zero*)
 * Reimplemnted numeric separators (*stage 3*)
 * Improved performance


src/masks.ts
// Simple immutable bit set passed around as arguments
export const enum Context {
    None                   = 0,
    Module                 = 1 << 0,   // If node was parsed in module code
    Strict                 = 1 << 1,   // If node was parsed in strict mode
    AllowIn                = 1 << 2,   // If node was parsed in a context where 'in-expressions' are allowed
    Arrow                  = 1 << 3,   // If node was parsed in a arrow context as plain identifier
    Yield                  = 1 << 4,   // If node was parsed in the 'yield' context created when parsing a generator
    Await                  = 1 << 5,   // If node was parsed in the 'await' context created when parsing an async function
    InParenthesis          = 1 << 6,   // If node was parsed in a parenthesized expression context
    InParameter            = 1 << 7,   // If node was parsed in the formal list of an function
    InAsyncArgs            = 1 << 8,   // If node was parsed in the formal list of an async arrow function
    InArrowParameterList   = 1 << 9,   // If node was parsed in the formal list of an arrow function
    Declaration            = 1 << 10,
    ExistingScope          = 1 << 11,  // If node was parsed in a context where the scope shouldn't be undefined
    AnnexB                 = 1 << 12,  // If node was parsed in the 'if statement' with the AnnexB semtantic
    Import                 = 1 << 13,  // Module ( strict and non-strict)
    Optional               = 1 << 14,  // Optional identifier for export of either anonymous class or function declaration
    Method                 = 1 << 15,  // If node was parsed in a object / class method definition context
    HasConstructor         = 1 << 16,  // If node was parsed inside Class and allow super
    AllowConstructor       = 1 << 17,  // Class declaration / expression
    ForStatement           = 1 << 18,  // If node was parsed in a for / for - in / for -of context
    Template               = 1 << 19,  //
    TaggedTemplate         = 1 << 20,  //
    Statement              = 1 << 21,  //
    ClassFields            = 1 << 22,  // If node was parsed in a class fields context
    Expression             = 1 << 23,
    Pattern                = 1 << 24,
    ValidateEscape         = 1 << 25,  // If node was parsed in a context where escaped keywords are forbidden
    StrictReserved         = 1 << 26,  // Variable declaration
    TopLevel               = 1 << 27,  // Variable declaration
    Let                    = 1 << 28,  // Variable declaration
    Const                  = 1 << 29,  // Variable declaration

    // An Lexical declaration can be either 'const¨' or 'let
    Lexical = Let | Const,

    // Duplicate formal params in either strict mode or async / await in sloppy mode
    MarkAsParamDuplicate = Strict | Yield | Await | Pattern
}

// Mutable parser flags
export const enum Flags {
    None                         = 0,
    LineTerminator               = 1 << 0,
    ExtendedUnicodeEscape        = 1 << 1, // If node has any escaped unicode sequences (escaped characters in keywords).
    InFunctionBody               = 1 << 2, // If node was parsed in a function body
    AllowCall                    = 1 << 3, // If node was parsed in a context where call should be allowed
    AllowBreak                   = 1 << 4,
    AllowContinue                = 1 << 5,
    HasPrototype                 = 1 << 6,
    Yield                        = 1 << 7,  // Used if we have seen a 'yield' token. E.g. in arrow formal param list
    Await                        = 1 << 8,  // Used if we have seen a 'await' token. E.g. in async arrow formal param list
    Rest                         = 1 << 9,
    Binding                      = 1 << 10, // Used if an "identifier" are used in binding position in strict mode
    DirectivePrologue            = 1 << 11, // Only used if we have seen a "use strict"; directive
    BigInt                       = 1 << 12, // e.g. `100n`
    SimpleParameterList          = 1 << 13,
    ParenthesizedPattern         = 1 << 14,
    Octal                        = 1 << 15, // e.g. `0777`
    ContainsSeparator            = 1 << 16,
    HasProtoField                = 1 << 17,
    HasDuplicateProtoField       = 1 << 18,
    StrictDirective              = 1 << 19,

    /* Options */
    OptionsTolerant              = 1 << 20,
    OptionsRanges                = 1 << 21, // Enable / disable "ranges"
    OptionsLoc                   = 1 << 22, // Enable / disable location tracking on the node
    OptionsSource                = 1 << 23,
    OptionsJSX                   = 1 << 24, // Enable / disable JSX extension
    OptionsRaw                   = 1 << 25, // Enable / disable "raw" property on the node
    OptionsNext                  = 1 << 26, // Enable / disable Stage 3 proposals
    OptionsGlobalReturn          = 1 << 27, // Allow return statement in global scope,
    OptionsSourceType            = 1 << 28, // Allow legacy method for settings sourceType - e.g. 'module' or 'script'
    OptionsComment               = 1 << 29,
    OptionsAttachComment         = 1 << 30,

    Comments = OptionsAttachComment | OptionsComment
}

// Flags used in parenthesized to validate arrow formal list
export const enum ParenthesizedState {
    None            = 0,
    EvalOrArg       = 1 << 0, // If (async) arrow contains eval or agruments
    Await           = 1 << 1, // If async arrow contains 'await'
    Parenthesized   = 1 << 2, // Tracks invalid parenthesized pattern
    Trailing        = 1 << 3, // Tracks trailing commas
    Pattern         = 1 << 4,
    FutureReserved  = 1 << 5,
    Yield           = 1 << 6,
}

// Flags used by both object expression and class decl / expr
export const enum ObjectState {
    None            = 0,
    Yield           = 1 << 0,
    Async           = 1 << 1,
    Computed        = 1 << 2,
    Shorthand       = 1 << 3,
    Get             = 1 << 4,
    Set             = 1 << 5,
    Method          = 1 << 6,
    HasConstructor  = 1 << 7,
    Heritage        = 1 << 8,
    Static          = 1 << 9,
    Constructor     = 1 << 10,
    Super           = 1 << 11,
    Prototype       = 1 << 12,
    HasPrototype    = 1 << 13,
    Delete          = 1 << 14,
    Accessors = Get | Set,
    Invalid =   Prototype | Constructor | Static,
    Modifiers = Accessors | Method | Yield,
    Special = Accessors | Yield | Async
}

// A set of flags for  maintaining the internal state machine.
export const enum ScanState {
    None            = 0,
    LastIsCR        = 1 << 0,
    LineStart       = 1 << 1,
    NewLine         = 1 << 2,
}

export const enum NumericState {
    None                    = 0,
    Decimal                 = 1 << 0,
    DecimalWithLeadingZero  = 1 << 1,
    Hexadecimal             = 1 << 2,
    Octal                   = 1 << 3,
    ImplicitOctal           = 1 << 4,
    Binary                  = 1 << 5,
    Float                   = 1 << 6,
    AllowSeparator          = 1 << 7,
    ContainsSeparator       = 1 << 8,
    BigInt                  = 1 << 9,

    Hibo = Hexadecimal | ImplicitOctal | Binary | Octal
}

// Regular expression scanning
export const enum RegExpState {
    Empty = 0,
    Escape = 0x1,
    Class = 0x2,
}

// Spidermonkey values
export const enum RegexFlags {
    None = 0,
    IgnoreCase = 1 << 0,
    Global     = 1 << 1,
    Multiline  = 1 << 2,
    Unicode    = 1 << 3,
    Sticky     = 1 << 4,
    DotAll     = 1 << 5,
}

export const enum FieldState {
    None = 0,
    Method = 1 << 0,
    Scope = 1 << 1,
}

export const enum ScopeMasks {
    Shadowable = 0x1,
    NonShadowable = 0x2,
}

export const enum Escape {
    Empty       = -1,
    StrictOctal = -2,
    EightOrNine = -3,
    InvalidHex  = -4,
    OutOfRange  = -5,
    TemplateOctalLiteral = -6,
}